source/agora/consensus/EnrollmentManager.d

```diff
@@ -475,22 +475,19 @@ public class EnrollmentManager
         Params:
             preimage = will contain the PreImageInfo if exists
             height = current block height
+            expected_height = height based on time offset from Genesis time
 
         Returns:
             true if the pre-image exists
 
     ***************************************************************************/
 
-    public bool getNextPreimage (out PreImageInfo preimage, in Height height)
+    public bool getNextPreimage (out PreImageInfo preimage, in Height height,
+        in Height expected_height)
```

Comment:
The problem is that the ledger does not have access to the `clock`. Do you suggest passing the clock to the `Ledger`?

Reply:
Ah, I'm very sorry. I meant that a validator should determine which height of pre-image should be revealed because the code looks complicated in `getNextPreimage`.

Reply:
I meant that validators should determine which height of pre-image should be revealed on `onPreImageRevealTimer` because the code of getNextPreimage looks complicated.

Reply:
I will move the logic up into `Validator`.

Reply:
Updated

Reply:
I meant the function should be as it was. The change is not related to what this PR is to achieve, I think. I'm not sure this change is really what we need for the moment.

Reply:
This change is to help prevent slashing due to late blocks so seems related to me.